userspace/tudumvc/updating-the-agent.md

```diff
@@ -0,0 +1,544 @@
+    * If it is not a valid `id`, it returns the state unchanged but mesasges the user in dojo indicating that there is `"No such task at ID {<id.action>}`.
+    * If it _is_ valid, it uses [`del:by`](https://urbit.org/docs/reference/library/2i/#del-by) to remove that key-value pair from `tasks` in the state.
+
+Try it in dojo by entering `:tudumvc &tudumvc-action [%remove-task 2]`. You should see something like this (assuming you have a task at that key `id`):
+```hoon
+>   "Removed task [~ [label='test' done=%.n]]"
+> :tudumvc &tudumvc-action [%remove-task 2]
+>=
+```
+
+##### `%mark-complete`
+```hoon
+  %mark-complete
+?.  (~(has by tasks) id.action)
+  ~&  >>>  "No such task at ID {<id.action>}"
+  `state
+=/  task-text=@tU  label.+<:(~(get by tasks) id.action)
+=/  done-state=?  ?!  done.+>:(~(get by tasks) id.action)
+~&  >  "Task {<task-text>} marked {<done-state>}"
+`state(tasks (~(put by tasks) id.action [task-text done-state]))
+```
+`%mark-complete` marks an existing task from the state `tasks` map as complete. It takes an argument of an `id=@ud` which should be one of the existing tasks `id`s in the map.
+
+This works almost identically to `%remove-task`, except that it only changes the `done`ness of a given task using `put:by` to replace the existing `done` value with the _opposite_ of the current `done` value. In the TodoMVC app, you can click the done button to mark an incomplete task as complete, and also click the done button of a _complete_ task to mark it _incomplete_. Allowing our app to automatically alternate between those solves for this functionality.
+
+`%tudumvc` performs this alternation of completeness function by creating a face of `done-state` and setting it equal to the _opposite_ of the current `done` state, using [`?!`](https://urbit.org/docs/reference/hoon-expressions/rune/wut/#wutzap) only _after_ we've confirmed that the `id` key actually exists.
+
+Try it in dojo a few times, using something like `:tudumvc &tudumvc-action [%mark-complete 1]`:
+```hoon
+>   "Task 'example task' marked %.y"
+> :tudumvc &tudumvc-action [%mark-complete 1]
+>=
+>   "Task 'example task' marked %.n"
+> :tudumvc &tudumvc-action [%mark-complete 1]
+>=
+```
+
+##### `%edit-task`
+We're not going to tell you how `%edit-task` works - in fact, that's one of your exercises this part of the guide. We can tell you it works rather similarly to the ones above, and that it takes two arguments:
+* An `id`.
+* An updated `label`.
+
+## Homework {#homework}
+* Examine the available structures of [JSON in Hoon](https://github.com/urbit/urbit/blob/6bcbbf8f1a4756c195a324efcf9515b6f288f700/pkg/arvo/sys/lull.hoon#L40), found in `lull.hoon`.
+* Read through [`++  enjs`](https://github.com/urbit/urbit/blob/6bcbbf8f1a4756c195a324efcf9515b6f288f700/pkg/arvo/sys/zuse.hoon#L3263) in `zuse.hoon`.
```

Comment:
Generally we remove spaces in gate names so `++enjs` if you want to have the ++ at the start.

Reply:
This is done in a number of places in a lot of files.

Reply:
Oh - and as part of the revision to tutorial, I'm taking out the side by side, preferring just code blocks of the updated content per "chapter" in the tutorial.  If I find any need for side-by-side again, I will use the tabbed version as suggested above.

Reply:
It's just not the same as it's done on the rest of the docs is my only concern.

Reply:
@baudtack Pushed the additional changes throughout - let me know if that style works better for the site.  Thank you again for your patience while I worked through that.